lib/translator/inquiry.cc

```diff
@@ -0,0 +1,253 @@
+
+    // PAGE LENGTH should be set to 30 indicating the page length in bytes.
+    result.page_length = kV1SerialLen;
+
+    // PRODUCT SERIAL NUMBER should be formed as follows:
+    // Bits 239:80 20 bytes of Serial Number (bytes 23:04 of Identify
+    // Controller data structure)
+    static_assert(sizeof(identify_controller_data.sn) == 20);
+    memcpy(&product_serial_number, identify_controller_data.sn,
+           sizeof(identify_controller_data.sn));
+
+    // Bits 79:72 ASCII representation of “_"
+    product_serial_number[sizeof(identify_controller_data.sn)] = '_';
+
+    // Bits 71:08 ASCII representation of 32 bit Namespace Identifier (NSID)
+    sprintf(&product_serial_number[sizeof(identify_controller_data.sn) + 1],
+            "%08x", nsid);
+
+    // Bits 07:00 ASCII representation of “."
+    product_serial_number[kV1SerialLen - 1] = '.';
+  }
+
+  memcpy(buffer.data(), &result, sizeof(result));
+  memcpy(&buffer[sizeof(result)], &product_serial_number, result.page_length);
+}
+
+StatusCode InquiryToNvme(absl::Span<const uint8_t> scsi_cmd,
+                         nvme_defs::GenericQueueEntryCmd &identify_ns,
+                         nvme_defs::GenericQueueEntryCmd &identify_ctrl,
+                         uint32_t &alloc_len) {
+  scsi_defs::InquiryCommand cmd =
+      *reinterpret_cast<const scsi_defs::InquiryCommand *>(&scsi_cmd[1]);
+  alloc_len = cmd.allocation_length;
+  // Allocate prp & assign to command
+  void *prp_ptr = AllocPages(1);
+  if (prp_ptr == nullptr) return StatusCode::kFailure;
+  uint64_t prp = reinterpret_cast<uint64_t>(prp_ptr);
+
+  identify_ns = nvme_defs::GenericQueueEntryCmd{
+      .opc = static_cast<uint8_t>(nvme_defs::AdminOpcode::kIdentify),
+  };
+  identify_ns.dptr.prp.prp1 = prp;
+  identify_ns.cdw[0] = 0x0;  // cns val
+
+  void *prp_ptr2 = AllocPages(1);
+  if (prp_ptr2 == nullptr) return StatusCode::kFailure;
+  uint64_t prp2 = reinterpret_cast<uint64_t>(prp_ptr2);
+
+  identify_ctrl = nvme_defs::GenericQueueEntryCmd{
+      .opc = static_cast<uint8_t>(nvme_defs::AdminOpcode::kIdentify),
+  };
+  identify_ctrl.dptr.prp.prp1 = prp2;
+  identify_ctrl.cdw[0] = 0x1;  // cns val
+  return StatusCode::kFailure;
+}
+
+// Main logic engine for the Inquiry command
+StatusCode InquiryToScsi(
+    absl::Span<const uint8_t> scsi_cmd, absl::Span<uint8_t> buffer,
+    absl::Span<const nvme_defs::GenericQueueEntryCmd> nvme_cmds) {
+  scsi_defs::InquiryCommand cmd =
+      *reinterpret_cast<const scsi_defs::InquiryCommand *>(&scsi_cmd[1]);
+
+  const nvme_defs::IdentifyControllerData identify_controller_data =
+      *reinterpret_cast<const nvme_defs::IdentifyControllerData *>(
+          nvme_cmds[0].dptr.prp.prp1);
+  const nvme_defs::IdentifyNamespace identify_namespace_data =
+      *reinterpret_cast<const nvme_defs::IdentifyNamespace *>(
+          nvme_cmds[1].dptr.prp.prp1);
+
+  // TODO: Verify if nsid should come from Controller or Namespace commands
+  uint32_t nsid = nvme_cmds[0].nsid;
```

Comment:
NSID is not set for IdentifyController so getting it from IdentifyNamespace is appropriate (addressing TODO)